### ISSUE TITLE
CLIP: Loss in implementation vs. in paper

### ISSUE THREAD

**Issue by iceychris** [lucidrains/DALLE-pytorch#32]
Hey!

In your implementation of CLIP, `F.cross_entropy` is called once with an implicit `F.log_softmax`
being computed along dimension `1` (because [`F.cross_entropy` internally calls `F.log_softmax(input, dim=1)`](https://pytorch.org/docs/stable/_modules/torch/nn/functional.html#cross_entropy)):

https://github.com/lucidrains/DALLE-pytorch/blob/9a335d79283e0a49e5b7e572da3af684215e479c/dalle_pytorch/dalle_pytorch.py#L236-L239

Whereas in the [CLIP-Paper](https://cdn.openai.com/papers/Learning_Transferable_Visual_Models_From_Natural_Language.pdf), they compute the cross-entropy-loss on both dimension `0` and `1` and then take an average:

![paper-loss](https://user-images.githubusercontent.com/12550267/106890633-e75d5a00-66e9-11eb-81c1-5de0a490010b.png)

So unless I am missing something here, one possible solution would just be something like

```python
sim = einsum('i d, j d -> i j', text_latents, image_latents) * temp 
labels = torch.arange(b, device = device) 
loss_t = F.cross_entropy(sim, labels) 
loss_i = F.cross_entropy(sim.T, labels) 
return (loss_t + loss_i) / 2. 
```

which, *in this case*, with `sim` being of shape `(n, n)`, should work?

**Comment by lucidrains** [comment]
@iceychris Hi Christian! You are exactly correct, making the change now!

**Comment by lucidrains** [comment]
done at https://github.com/lucidrains/DALLE-pytorch/releases/tag/0.0.49

**Comment by CDitzel** [comment]
this whole contrastive loss function topic is quite controversial, as I take it. I reference here a discussion over at [OpenAIs github page](https://github.com/openai/CLIP/issues/25) where even Alec joins in.

**Comment by DebajyotiS** [comment]
Am I missing something obvious here, or should the temperature be divided instead of multiplied? The loss in the paper reads exp(s/t). What is the functional form used in the functional.cross_entropy?
